### ISSUE TITLE
Conv-FNO

### ISSUE THREAD

**Issue by ach14012022** [neuraloperator/neuraloperator#97]
@scaomath, @zongyi-li 
Hi everyone,
Any idea how we can replace MLP in FNO with convolutional layer for 2D Darcy flow?
https://github.com/neural-operator/fourier_neural_operator/blob/master/fourier_2d.py
Tnx a lot.

**Comment by ach14012022** [comment]
Thanks for your response.
I mean in the main code lines 93-96 MLP is used. In order to improve the accuracy, it seems that it is possible to replace MLP with Convolutional layers e.g. U-FNO architecture. How we can do that?

**Comment by zongyi-li** [comment]
You can start with a large kernel size, like 3x3. Feel free to replace it with a UNet block or any other architecture you like.

**Comment by scaomath** [comment]
If you mean

https://github.com/neural-operator/fourier_neural_operator/blob/af93f781d5e013f8ba5c52baa547f2ada304ffb0/fourier_2d.py#L116-L120

It seems @zongyi-li tweaked the base code a bit. The basic magical block in the original FNO is

```python
x1 = self.conv1(x)
x2 = self.w1(x)
x = x1 + x2 
x = F.gelu(x) 
```
where `conv1` is the spectral convolution and `w1` is a universal approximator that has a `tensor2tensor` input output format, you can change it to convolution, as long as it gets a `(bsz, C, n, n)` tensor and outputs another tensor of the same size. I tried myself before using 3x3 conv, the improvement is very minor on the current Darcy dataset.

**Comment by ach14012022** [comment]
Thanks for your response. Would you please share your code? I want to try it on another set of 2D Darcy data.

**Comment by ach14012022** [comment]
@scaomath @zongyi-li 
Can you please explain the idea behind adding mlp?

**Comment by scaomath** [comment]
> Thanks for your response. Would you please share your code? I want to try it on another set of 2D Darcy data.

You can try changing the MLP in `self.w1` to `nn.Conv2d`.

> @scaomath @zongyi-li Can you please explain the idea behind adding mlp?

Please check the paragraph right before Section 2 in the original FNO paper page 3: https://openreview.net/pdf?id=c8P9NQVtmnO
Please also see another take here on page 8 of the Galerkin transformer paper on the role of MLP: https://openreview.net/pdf?id=ssohLcmn4-r

**Comment by ach14012022** [comment]
> n

I found I used the previous code, in which there was no self.mlp. By running the new version of code, the performance of FNO on the testing data increased (my own data).
For the role of MLP, I think you mean this: Despite the fact that the FFNs, when being viewed as a class of functions, bear no linear structure within, the basis functions produced this way act as a building block to characterize a linear space for a learnable projection. Am I right? Meantime, please explain in a simpler way, if possible.

**Comment by scaomath** [comment]
> > n
> 
> I found I used the previous code, in which there was no self.mlp. By running the new version of code, the performance of FNO on the testing data increased (my own data). For the role of MLP, I think you mean this: Despite the fact that the FFNs, when being viewed as a class of functions, bear no linear structure within, the basis functions produced this way act as a building block to characterize a linear space for a learnable projection. Am I right? Meantime, please explain in a simpler way, if possible.

If you treat each channel (column) as a "basis", then $P(u + \text{MLP}(u))$ is like a projection, where $P$'s size depends only on the channels (not the size of the discretization like DeepONets).

**Comment by JeanKossaifi** [comment]
We also added an option to have an MLP (doing channel mixing) in addition to the global conv in the latest version. 
Closing, feel free to reopen if you have more questions.
